### PR TITLE
Add Support for Data Annotations

### LINKED ISSUE #107: Convention-based model builder support for Data Annotations
We should try to support at least the very basic ones, e.g. KeyAttribute for alpha, provided that the DataAnnotations build is ready for K.
Attributes to be implemented:
- [x] Key
- [x] Required
- [x] MaxLength
- [x] NotMapped
- [x] ConcurrencyCheck
- [x] TimeStamp
- [x] Table
- [x] Column
- [x] DatabaseGenerated
- [x] StringLength

Relationship
- [x] ForeignKey
- [x] InverseProperty
- [x] Required on Navigations

Not supported yet
- ComplexType

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #107 
Attributes added
* Table
* Column
* Key
* ConcurrencyCheck
* DatabaseGenerated
* Required

### CODE REVIEW

**Comment by smitpatel** [comment]
Will file following issue when this is accepted.
* Implement ColumnOrder fully
* StoreGeneratedPattern or DatabaseGenerated enum
* The remaining attributes to be added (can add them as a part of this PR also)

**Comment by rowanmiller** [comment]
@smitpatel rather than filing new issues, can you just add them as a checklist on #107. I think it makes sense just to leave that active until we have actually fully implemented data annotations.

File: src/EntityFramework.Core/EntityFramework.Core.csproj [hunk 1]
```diff
@@ -141,12 +142,18 @@
     <Compile Include="DbContextOptionsBuilder`.cs" />
     <Compile Include="Metadata\Builders\ReferenceReferenceBuilder`.cs" />
     <Compile Include="Metadata\CoreAnnotationNames.cs" />
+    <Compile Include="Metadata\ModelConventions\AttributeConventions\EntityTypeAttributes\EntityTypeAttributeConvention.cs" />
+    <Compile Include="Metadata\ModelConventions\AttributeConventions\PropertyAttributes\KeyAttributeConvention.cs" />
+    <Compile Include="Metadata\ModelConventions\AttributeConventions\PropertyAttributes\DatabaseGeneratedAttributeConvention.cs" />
+    <Compile Include="Metadata\ModelConventions\AttributeConventions\PropertyAttributes\ConcurrencyCheckAttributeConvention.cs" />
+    <Compile Include="Metadata\ModelConventions\AttributeConventions\PropertyAttributes\RequiredAttributeConvention.cs" />
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: No need to add subfolders, just have all the conventions under ModelConventions.

File: src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+            var newKey = entityTypeBuilder.PrimaryKey(properties, ConfigurationSource.Convention);
+            if (newKey != null)
+            {
+                entityTypeBuilder.PrimaryKey(properties, ConfigurationSource.DataAnnotation);
+                return;
+            }
+
+            newKey = entityTypeBuilder.PrimaryKey(properties, ConfigurationSource.DataAnnotation);
+            if (newKey != null)
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs]: This check and the .PrimaryKey call are unnecessary, just make the final call

File: src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            {
+                properties = currentKey.Properties.Select(p => p.Name).ToList();
+                properties.Add(propertyBuilder.Metadata.Name);
+                entityTypeBuilder.PrimaryKey(properties, ConfigurationSource.DataAnnotation);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs]: Sort `properties`, even though we will probably throw later.

File: src/EntityFramework.Core/Metadata/ModelConventions/ConventionDispatcher.cs [hunk 4]
```diff
@@ -108,5 +108,21 @@ public virtual InternalModelBuilder InitializingModel([NotNull] InternalModelBui
 
             return modelBuilder;
         }
+
+        public virtual InternalModelBuilder ValidatingModel([NotNull] InternalModelBuilder modelBuilder)
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/ModelConventions/ConventionDispatcher.cs]: Rename to OnModelBuilt to be consistent with other methods.

File: src/EntityFramework.Core/ModelBuilder.cs [hunk 5]
```diff
@@ -58,6 +58,13 @@ public ModelBuilder([NotNull] ConventionSet conventions, [NotNull] Model model)
             _builder = new InternalModelBuilder(model, conventions).Initialize();
         }
 
+        public virtual ModelBuilder Validate()
+        {
+            _builder = Builder.Validate();
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/ModelBuilder.cs]: It doesn't make a lot of sense replacing the model builder in a convention. `IModelConvention.Apply` should just return void

File: src/EntityFramework.Relational/Metadata/ModelConventions/AttributeConventions/EntityTypeAttributes/RelationalTableAttributeConvention.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    public class RelationalTableAttributeConvention : EntityTypeAttributeConvention<TableAttribute>
+    {
+        public override void Apply(InternalEntityTypeBuilder entityTypeBuilder, TableAttribute attribute)
+        {
+            Check.NotNull(entityTypeBuilder, nameof(entityTypeBuilder));
+            Check.NotNull(attribute, nameof(attribute));
+
+            if (!string.IsNullOrWhiteSpace(attribute.Schema))
+            {
+                entityTypeBuilder.Metadata.Relational().Schema = attribute.Schema;
```
Comment (AndriySvyryd) [inline src/EntityFramework.Relational/Metadata/ModelConventions/AttributeConventions/EntityTypeAttributes/RelationalTableAttributeConvention.cs]: The annotations need to be set through entityTypeBuilder.Annotation

File: src/EntityFramework.Relational/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/RelationalColumnAttributeConvention.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    public class RelationalColumnAttributeConvention : PropertyAttributeConvention<ColumnAttribute>
+    {
+        public override void Apply(InternalPropertyBuilder propertyBuilder, ColumnAttribute attribute)
+        {
+            Check.NotNull(propertyBuilder, nameof(propertyBuilder));
+            Check.NotNull(attribute, nameof(attribute));
+
+            if (!string.IsNullOrWhiteSpace(attribute.Name))
+            {
+                propertyBuilder.Metadata.Relational().Column = attribute.Name;
```
Comment (AndriySvyryd) [inline src/EntityFramework.Relational/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/RelationalColumnAttributeConvention.cs]: The annotations need to be set through propertyBuilder.Annotation

File: src/EntityFramework.Relational/RelationalEntityServicesBuilderExtensions.cs [hunk 8]
```diff
@@ -34,6 +34,7 @@ public static EntityFrameworkServicesBuilder AddRelational([NotNull] this Entity
                 .AddSingleton<TypedValueBufferFactoryFactory>()
                 .AddSingleton<IMigrationModelFactory, MigrationModelFactory>()
                 .AddSingleton<RelationalModelValidator>()
+                .AddSingleton<RelationalConventionSetBuilder>()
```
Comment (AndriySvyryd) [inline src/EntityFramework.Relational/RelationalEntityServicesBuilderExtensions.cs]: You don't need to register RelationalConventionSetBuilder, the providers need to use it or a derived one.

File: test/EntityFramework.Core.Tests/Metadata/ModelConventions/PropertyAttributeConventionTest.cs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        [Fact]
+        public void KeyAttribute_does_not_throw_when_setting_composite_primary_key_if_fluent_api_used()
+        {
+            var model = new MyContext().Model;
+
+            Assert.Equal(2, model.GetEntityType(typeof(B)).GetPrimaryKey().Properties.Count);
+            Assert.Equal("MyPrimaryKey", model.GetEntityType(typeof(B)).GetPrimaryKey().Properties[0].Name);
+            Assert.Equal("Id", model.GetEntityType(typeof(B)).GetPrimaryKey().Properties[1].Name);
+        }
+
```
Comment (AndriySvyryd) [inline test/EntityFramework.Core.Tests/Metadata/ModelConventions/PropertyAttributeConventionTest.cs]: Add tests that ensure these conventions override configuration from non-data annotation conventions and don't override explicit configuration

File: test/EntityFramework.Relational.Tests/Metadata/RelationalEntityTypeAttributeConventionTest.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+        [Fact]
+        public void TableAttribute_set_table_name_and_schema_for_entity_type()
+        {
+            var modelBuilder = new ModelBuilder(new RelationalConventionSetBuilder().AddConventions(new CoreConventionSetBuilder().CreateConventionSet()));
+            var entityType = modelBuilder.Entity<A>();
+
+            Assert.Equal("MyTable", entityType.Metadata.Relational().Table);
+            Assert.Equal("MySchema", entityType.Metadata.Relational().Schema);
+        }
```
Comment (AndriySvyryd) [inline test/EntityFramework.Relational.Tests/Metadata/RelationalEntityTypeAttributeConventionTest.cs]: Same comment as above

File: src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs (hunk 3 above)
Comment (smitpatel) [inline src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs]: Sort on alphabetical order?

**Comment by AndriySvyryd** [comment]
Add a functional test that uses some data annotations and runs for every provider.
Also add a similar test for relational data annotations.

File: src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/ModelConventions/AttributeConventions/PropertyAttributes/KeyAttributeConvention.cs]: Yes

**Comment by smitpatel** [comment]
Updated with feedback.

File: test/EntityFramework.Relational.FunctionalTests/DataAnnotationRelationalTestBase.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        where TTestStore : TestStore
+        where TFixture : DataAnnotationFixtureBase<TTestStore>, new()
+    {
+        protected DataAnnotationRelationalTestBase(TFixture fixture)
+            : base(fixture)
+        {
+        }
+
+        public override void RequiredAttribute_throws_while_inserting_null_value()
+        {
```
Comment (AndriySvyryd) [inline test/EntityFramework.Relational.FunctionalTests/DataAnnotationRelationalTestBase.cs]: This method is redundant and can be removed

File: test/EntityFramework.Relational.FunctionalTests/DataAnnotationRelationalTestBase.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+                    innerContext.SaveChanges();
+                }
+
+                Assert.Equal(Strings.UpdateConcurrencyException(1, 0),
+                    Assert.Throws<DbUpdateConcurrencyException>(() => context.SaveChanges()).Message);
+            }
+        }
+
+        public override void DatabaseGeneratedAttribute_autogenerates_values_when_set_to_identity()
```
Comment (AndriySvyryd) [inline test/EntityFramework.Relational.FunctionalTests/DataAnnotationRelationalTestBase.cs]: This method is redundant and can be removed

**Comment by AndriySvyryd** [comment]
:shipit:

**Pull request merged**